src/Components/Aspire.StackExchange.Redis.DistributedCaching/AspireRedisDistributedCacheExtensions.cs

```diff
@@ -106,4 +125,54 @@ private static void AddRedisDistributedCacheCore(this IHostApplicationBuilder bu
             .Configure(configureRedisOptions);
     }
+
+    /// <summary>
+    /// Configures the Redis client to provide a keyed distributed caching service through <see cref="IDistributedCache"/> using <paramref name="name"/> as the key.
+    /// </summary>
+    /// <param name="builder">The <see cref="AspireRedisClientBuilder"/> to configure.</param>
+    /// <param name="name">The name which is used as the <see cref="ServiceDescriptor.ServiceKey"/> of the service.</param>
+    /// <param name="configureOptions">An optional method that can be used for customizing the <see cref="RedisCacheOptions"/>.</param>
+    /// <returns>The <see cref="AspireRedisClientBuilder"/> for method chaining.</returns>
+    /// <example>
+    /// The following example creates keyed IDistributedCache service using the "myCache" key for a Redis client connection named "redis".
+    /// <code lang="csharp">
+    /// var builder = WebApplication.CreateBuilder(args);
+    ///
+    /// builder.AddRedisClientBuilder("redis")
+    ///        .WithKeyedDistributedCache("myCache", options => options.InstanceName = "myCache");
+    /// </code>
+    /// The created IDistributedCache service can then be resolved using the "myCache" key:
+    /// <code lang="csharp">
+    /// IServiceProvider serviceProvider = builder.Services.BuildServiceProvider();
+    ///
+    /// var cache = serviceProvider.GetRequiredKeyedService&lt;IDistributedCache&gt;("myCache");
+    /// </code>
+    /// </example>
+    public static AspireRedisClientBuilder WithKeyedDistributedCache(this AspireRedisClientBuilder builder, string name, Action<RedisCacheOptions>? configureOptions = null)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        if (configureOptions is not null)
+        {
+            // note that AddStackExchangeRedisCache doesn't support named RedisCacheOptions options, so we need to register the named options ourselves
+            builder.HostBuilder.Services
+                .AddOptions<RedisCacheOptions>(name)
+                .Configure(configureOptions);
+        }
+
+        builder.HostBuilder.Services.AddKeyedSingleton<IDistributedCache>(name, (sp, key) =>
+        {
+            var options = sp.GetRequiredService<IOptionsMonitor<RedisCacheOptions>>().Get((string?)key);
+
+            options.ConnectionMultiplexerFactory = () => Task.FromResult(sp.GetRequiredKeyedService<IConnectionMultiplexer>(builder.ServiceKey));
+
+            // AddStackExchangeRedisCache only supports unkeyed IDistributedCache, so we need to create the RedisCache instance ourselves.
+            // As of .NET 10, the drawback to this approach is that it doesn't use the internal RedisCacheImpl class. Which means:
+            // - The RedisCache won't log appropriately (but it only logs in one place - where it is unable to add library name suffix.
+            // - If HybridCache is being used, it won't add the 'HC' library name suffix to the connection.
```

Comment:
This happens because we don't end up utilizing [this IsHybridCache check](https://github.com/dotnet/aspnetcore/blob/8ee12ef7a2c179f3d7c7da5ab33d76d652042d0b/src/Caching/StackExchangeRedis/src/RedisCacheImpl.cs#L16), correct?

I think it's only used for the `HC` suffix right now, as you mentioned, but we might want to be careful about coding being introduced that uses that check that might cause more issues.

Do we have a plan on how to fix this permanently for .NET 11 and beyond?

Reply:
> This happens because we don't end up utilizing [this IsHybridCache check](https://github.com/dotnet/aspnetcore/blob/8ee12ef7a2c179f3d7c7da5ab33d76d652042d0b/src/Caching/StackExchangeRedis/src/RedisCacheImpl.cs#L16), correct?

Correct. Since we don't get a derived `RedisCacheImpl` instance created, this check doesn't run.

> Do we have a plan on how to fix this permanently for .NET 11 and beyond?

I opened [Add support for keyed IDistributedCache in AddStackExchangeRedisCache (dotnet/aspnetcore#63700)](https://github.com/dotnet/aspnetcore/issues/63700)